examples/step-69/step-69.cc

```diff
@@ -2842,7 +2840,28 @@ namespace Step69
     // the <code>this</code> pointer as well as most of the arguments of
     // the output function by value so that we have access to them inside
     // the lambda function.
-    const auto output_worker = [this, name, t, cycle, checkpoint, data_out]() {
+    //
+    // The last capture argument of the lambda function, `data_out_copy`
+    // in essence creates a local variable inside the lambda function into
+    // which we "move" the `data_out` variable from above. The way this works
+    // is that we create a `std::unique_ptr` above that points to the DataOut
+    // object. But we have no use for it any more after this point: We really
+    // just want to move ownership from the current function to the lambda
+    // function implemented in the following few lines. We could have done
+    // this by using a `std::shared_ptr` above and giving the lambda function
+    // a copy of that shared pointer; once the current function returns (but
+    // maybe while the lambda function is still running), our local shared
+    // pointer would go out of scope and stop pointing at the actual
+    // object, at which point the lambda function simply becomes the sole
+    // owner. But using the `std::unique_ptr` is conceptually cleaner as it
+    // makes it clear that the current function's `data_out` variable isn't
+    // even pointing to the object any more.
+    const auto output_worker = [this,
+                                name,
+                                t,
+                                cycle,
+                                checkpoint,
+                                data_out_copy = std::move(data_out)]() {
```

Comment:
It seems that gcc-5.4.0 is tripping over the move semantics.
```
/jenkins/workspace/dealii/PR-10397/examples/step-69/step-69.cc:2864:68: note: ‘Step69::MainLoop<dim>::output(const vector_type&, const string&, double, unsigned int, bool) [with int dim = 2; Step69::MainLoop<dim>::vector_type = std::array<dealii::LinearAlgebra::distributed::Vector<double>, 4ul>; std::__cxx11::string = std::__cxx11::basic_string<char>]::<lambda()>::<lambda>(const Step69::MainLoop<dim>::output(const vector_type&, const string&, double, unsigned int, bool) [with int dim = 2; Step69::MainLoop<dim>::vector_type = std::array<dealii::LinearAlgebra::distributed::Vector<double>, 4ul>; std::__cxx11::string = std::__cxx11::basic_string<char>]::<lambda()>&)’ is implicitly deleted because the default definition would be ill-formed:
                                 data_out_copy = std::move(data_out)]() {
                                                                    ^
/jenkins/workspace/dealii/PR-10397/examples/step-69/step-69.cc:2864:68: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = dealii::DataOut<2, dealii::DoFHandler<2, 2> >; _Dp = std::default_delete<dealii::DataOut<2, dealii::DoFHandler<2, 2> > >]’
In file included from /usr/include/c++/5/bits/locale_conv.h:41:0,
                 from /usr/include/c++/5/locale:43,
                 from /usr/include/c++/5/iomanip:43,
                 from /home/dealii/libs/trilinos-release-12-10-1/include/Teuchos_ConfigDefs.hpp:89,
                 from /home/dealii/libs/trilinos-release-12-10-1/include/Teuchos_PromotionTraits.hpp:45,
                 from /home/dealii/libs/trilinos-release-12-10-1/include/Sacado_Fad_DFadTraits.hpp:142,
                 from /home/dealii/libs/trilinos-release-12-10-1/include/Sacado.hpp:51,
                 from /jenkins/workspace/dealii/PR-10397/include/deal.II/base/numbers.h:151,
                 from include/deal.II/base/config.h:453,
                 from /jenkins/workspace/dealii/PR-10397/include/deal.II/base/conditional_ostream.h:19,
                 from /jenkins/workspace/dealii/PR-10397/examples/step-69/step-69.cc:41:
/usr/include/c++/5/bits/unique_ptr.h:356:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^
```

Reply:
Ugh. I'm pretty sure my code is right, though.

Also, how did they get that wrong? That was specifically one of the two points I can think of where it makes sense to use these sorts of initializers in lambda captures (capturing member variables of larger objects being the other).

Reply:
I mean my code is straight out of the book: https://stackoverflow.com/questions/25408190/lambda-captures-c14

Reply:
The code is absolutely fine - gcc-5.4.0 is broken. I have no good idea how to proceed. (I very much prefer the explicit move semantics of the unique_ptr as done in your pull request.)

Reply:
(I actually use precisely what you have in your pull request in our higher-order code together with an explicit `.reset()` in order to reliably free up the memory after writeout.)